### DISCUSSION TITLE
Skidl Jupyter book

### DISCUSSION THREAD

**Discussion by shanemmattner** [xesscorp/skidl#153] [Ideas]
@devbisme what do you think of taking the [skidl documentation](https://devbisme.github.io/skidl/) and putting it into a [Jupyter book](https://jupyterbook.org/en/stable/intro.html)?  I've been helping to write a Jupyter book for [Plotly Dash](https://open-resources.github.io/dash_curriculum/intro.html) and it turned out really well.

**Comment by devbisme** [comment]
That's a good idea!

I previously made a notebook to show how to use skidl to do spice
simulations. I've also been thinking about using notebooks to explain the
list of "nifty fifty" transistor circuits.

**Reply by shanemmattner** [reply]
Great!  Here's a [link](https://shanemmattner.github.io/skidl_jupyter_book/intro.html) to the book.  I'll plan on copying over everything you have in this [SKiDL page](https://devbisme.github.io/skidl/).  Let me know if there's anything you'd like changed/updated.

**Comment by xesscorp** [comment]
Looks good. Let's just go with whatever is in the current master branch
documentation.

**Comment by shanemmattner** [comment]
@devbisme all done.  Take a look and let me know what you think.  I'm going to add the [spice sim intro](https://github.com/devbisme/skidl/blob/master/examples/spice-sim-intro/spice-sim-intro.ipynb) chapter as well

[https://shanemmattner.github.io/skidl_jupyter_book/intro.html](https://shanemmattner.github.io/skidl_jupyter_book/intro.html)

**Comment by devbisme** [comment]
Wow, that looks really, *really* good! I like having the chapters listed in the left column and the subsections appearing in the right column. And there's even a search feature!

How easy is it to add executable code cells to the book so the reader can try things? This would be great with the SPICE examples, but requires interfacing to an ngspice executable which might be tricky.

**Reply by shanemmattner** [reply]
It is possible and I've got it kinda working:  Go to [chapter 1](https://shanemmattner.github.io/skidl_jupyter_book/chapter1/02_basic_usage.html) and click the rocket ship in the upper right corner and click `Live Code`.

What's interesting is that when the book loads we see the correct value being printed out:
![pre-compile](https://user-images.githubusercontent.com/22162612/179447901-cfdee95e-1c20-4e9b-9552-9532677b207b.png)

However, when you do `Live Code` and try to run the cell I get errors for KiCad being missing:

```
Requirement already satisfied: skidl in /srv/conda/envs/notebook/lib/python3.7/site-packages (1.1.0)
Requirement already satisfied: graphviz in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl) (0.20)
Requirement already satisfied: kinparse>=0.1.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl) (1.1.0)
Requirement already satisfied: future>=0.15.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl) (0.18.2)
Requirement already satisfied: kinet2pcb in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl) (1.0.1)
Requirement already satisfied: pyparsing>=2.1.1 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from kinparse>=0.1.0->skidl) (3.0.7)
Requirement already satisfied: hierplace in /srv/conda/envs/notebook/lib/python3.7/site-packages (from kinet2pcb->skidl) (1.0.0)
WARNING: KICAD_SYMBOL_DIR environment variable is missing, so the default KiCad symbol libraries won't be searched. @ [/home/jovyan/<frozen importlib._bootstrap_external>:728=>/home/jovyan/<frozen importlib._bootstrap>:219]
WARNING: Could not load KiCad schematic library "Device", falling back to backup library. @ [/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3524=>/tmp/ipykernel_384/2907924650.py:5]
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/tmp/ipykernel_384/2907924650.py in <module>
      3 from skidl import *
      4 
----> 5 resistor = Part('Device','R')
      6 resistor.value = '3K'
      7 resistor.value

~/skidl/part.py in __init__(self, lib, name, dest, tool, connections, part_defn, circuit, ref_prefix, ref, tag, pin_splitters, tool_version, **kwargs)
    187                         lib = skidl.load_backup_lib()
    188                         if not lib:
--> 189                             raise e
    190                     else:
    191                         raise e

~/skidl/part.py in __init__(self, lib, name, dest, tool, connections, part_defn, circuit, ref_prefix, ref, tag, pin_splitters, tool_version, **kwargs)
    177                 libname = lib
    178                 try:
--> 179                     lib = SchLib(filename=libname, tool=tool)
    180                 except FileNotFoundError as e:
    181                     if skidl.QUERY_BACKUP_LIB:

~/skidl/schlib.py in __init__(self, filename, tool, lib_section, **attribs)
     83             else:
     84                 load_func(
---> 85                     filename, skidl.lib_search_paths[tool], lib_section=lib_section
     86                 )
     87                 self.filename = filename

~/skidl/tools/kicad/kicad.py in load_sch_lib(self, filename, lib_search_paths_, lib_section)
     66     if not f:
     67         raise FileNotFoundError(
---> 68             "Unable to open KiCad Schematic Library File {}".format(filename)
     69         )
     70 

FileNotFoundError: Unable to open KiCad Schematic Library File Device
```

**Reply by shanemmattner** [reply]
What do you think of making SKiDL's default `KiCad Schematic Library` path be pointed towards the github KiCad Symbols Library [v5](https://github.com/KiCad/kicad-symbols) or [v6](https://gitlab.com/kicad/libraries/kicad-symbols)

Also, is there a way to set the `KiCad Schematic Library` path in a SKiDL script?  Either to a local or web path.

**Reply by devbisme** [reply]
I imagine when Jupyter Book compiles the document, it probably has access to your local KiCad library and than caches the results in the final HTML. But local files can't be accessed for security reasons when the Live Code button is pressed in the browser?

I'll work on getting SKiDL to use the library repositories on Gitlab. That's a good idea.

**Comment by devbisme** [comment]
I added the ability to read KiCad symbol libraries from the Gitlab remote repositories. To access V5 symbols, add the following to the list in `lib_search_paths[KICAD]`:
```
https://raw.githubusercontent.com/KiCad/kicad-symbols/master/
```
For V6, use:
```
https://gitlab.com/kicad/libraries/kicad-symbols/-/raw/master
```

**Reply by shanemmattner** [reply]
Looks like it's working.  I moved my KiCAD schematic files and tried to run my example code which gave an error:
```
(venv) [shanemattner@fedora-1 skidl]$ python ~/Desktop/esp32/skidl_esp32/esp32.py
WARNING: Could not load KiCad schematic library "RF_Module", falling back to backup library. @ [/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py:172=>/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py:54]
Traceback (most recent call last):
  File "/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py", line 172, in <module>
    esp32_circuit()
  File "/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py", line 54, in esp32_circuit
    esp32 = Part('RF_Module','ESP32-WROOM-32', footprint='RF_Module:ESP32-WROOM-32')
  File "/home/shanemattner/Desktop/skidl/skidl/part.py", line 189, in __init__
    raise e
  File "/home/shanemattner/Desktop/skidl/skidl/part.py", line 179, in __init__
    lib = SchLib(filename=libname, tool=tool)
  File "/home/shanemattner/Desktop/skidl/skidl/schlib.py", line 85, in __init__
    load_func(
  File "/home/shanemattner/Desktop/skidl/skidl/tools/kicad/kicad.py", line 85, in load_sch_lib
    raise FileNotFoundError(
FileNotFoundError: Unable to open KiCad Schematic Library File RF_Module
```

Then I added the line below and the netlist generated fine:
```python
lib_search_paths[KICAD].append("https://gitlab.com/kicad/libraries/kicad-symbols/-/raw/master")
```
SKiDL output:
```
(venv) [shanemattner@fedora-1 skidl]$ python ~/Desktop/esp32/skidl_esp32/esp32.py
WARNING: Found multiple parts matching DMN2041L. Selecting BSS138. @ [/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py:71=>/home/shanemattner/Desktop/esp32/skidl_esp32/esp32.py:40]

No errors or warnings found while generating netlist.
```

**Reply by shanemmattner** [reply]
Although I still can't get the example to run in the Jupyter book using this script:
```python
!pip install git+https://github.com/devbisme/skidl@development
!export KICAD_SYMBOL_DIR="https://github.com/shanemmattner/skidl_jupyter_book/tree/master/jupyter_book/symbols"
from skidl import *
lib_search_paths[KICAD].append("https://gitlab.com/kicad/libraries/kicad-symbols/-/raw/master")


resistor = Part('Device','R')
resistor.value = '1K' 
resistor.value
```
I get the same error as I have with `master`:
```
Collecting git+https://github.com/devbisme/skidl@development
  Cloning https://github.com/devbisme/skidl (to revision development) to /tmp/pip-req-build-qaaps29b
  Running command git clone --filter=blob:none --quiet https://github.com/devbisme/skidl /tmp/pip-req-build-qaaps29b
  Running command git checkout -b development --track origin/development
  Switched to a new branch 'development'
  Branch 'development' set up to track remote branch 'development' from 'origin'.
  Resolved https://github.com/devbisme/skidl to commit 6875bf697a469fcf5ee9321acab01d777e0ad2e8
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: kinet2pcb in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl==1.2.0) (1.0.1)
Requirement already satisfied: future>=0.15.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl==1.2.0) (0.18.2)
Requirement already satisfied: graphviz in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl==1.2.0) (0.20)
Requirement already satisfied: kinparse>=0.1.0 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from skidl==1.2.0) (1.1.0)
Requirement already satisfied: pyparsing>=2.1.1 in /srv/conda/envs/notebook/lib/python3.7/site-packages (from kinparse>=0.1.0->skidl==1.2.0) (3.0.7)
Requirement already satisfied: hierplace in /srv/conda/envs/notebook/lib/python3.7/site-packages (from kinet2pcb->skidl==1.2.0) (1.0.0)
WARNING: KICAD_SYMBOL_DIR environment variable is missing, so the default KiCad symbol libraries won't be searched. @ [/home/jovyan/<frozen importlib._bootstrap_external>:728=>/home/jovyan/<frozen importlib._bootstrap>:219]
WARNING: Could not load KiCad schematic library "Device", falling back to backup library. @ [/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3524=>/tmp/ipykernel_371/3966442685.py:7]
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/tmp/ipykernel_371/3966442685.py in <module>
      5 
      6 
----> 7 resistor = Part('Device','R')
      8 resistor.value = '1K'
      9 resistor.value

~/skidl/part.py in __init__(self, lib, name, dest, tool, connections, part_defn, circuit, ref_prefix, ref, tag, pin_splitters, tool_version, **kwargs)
    187                         lib = skidl.load_backup_lib()
    188                         if not lib:
--> 189                             raise e
    190                     else:
    191                         raise e

~/skidl/part.py in __init__(self, lib, name, dest, tool, connections, part_defn, circuit, ref_prefix, ref, tag, pin_splitters, tool_version, **kwargs)
    177                 libname = lib
    178                 try:
--> 179                     lib = SchLib(filename=libname, tool=tool)
    180                 except FileNotFoundError as e:
    181                     if skidl.QUERY_BACKUP_LIB:

~/skidl/schlib.py in __init__(self, filename, tool, lib_section, **attribs)
     83             else:
     84                 load_func(
---> 85                     filename, skidl.lib_search_paths[tool], lib_section=lib_section
     86                 )
     87                 self.filename = filename

~/skidl/tools/kicad/kicad.py in load_sch_lib(self, filename, lib_search_paths_, lib_section)
     66     if not f:
     67         raise FileNotFoundError(
---> 68             "Unable to open KiCad Schematic Library File {}".format(filename)
     69         )
     70 

FileNotFoundError: Unable to open KiCad Schematic Library File Device
```

**Reply by devbisme** [reply]
Yes, I'm getting this same error. Time to add some debugging statements...

**Reply by devbisme** [reply]
I made a `dev_debug` branch and added some `print` statements to see what is going on. I can see the debug output when I run in a local terminal, but none of these statements get triggered (no output) when I run the Jupyter Book page using Live Code. Then I ran the `master` branch locally and got the same output as we're seeing in Jupyter Book. It's almost like the Book is using the `master` branch even though I can clearly see it reference the commit hash for the `dev_debug` branch in the output.

**Comment by shanemmattner** [comment]
[This Google Colab script](https://colab.research.google.com/drive/1yVSVqZwZ1S4ZTO5UY4iVzxedFPRCtz7G?usp=sharing) can install the specific branch and works fine.  
```python
!pip install git+https://github.com/devbisme/skidl@development
from skidl import *
lib_search_paths[KICAD].append("https://gitlab.com/kicad/libraries/kicad-symbols/-/raw/master")


resistor = Part('Device','R', footprint='Resistor_SMD:R_0603_1608Metric')
resistor.value = '1K' 
resistor.value

generate_netlist(file_='test.net')
```
Console output:
```
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Collecting git+https://github.com/devbisme/skidl@development
  Cloning https://github.com/devbisme/skidl (to revision development) to /tmp/pip-req-build-0_my0gxw
  Running command git clone -q https://github.com/devbisme/skidl /tmp/pip-req-build-0_my0gxw
  Running command git checkout -b development --track origin/development
  Switched to a new branch 'development'
  Branch 'development' set up to track remote branch 'development' from 'origin'.
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: kinet2pcb in /usr/local/lib/python3.7/dist-packages (from skidl==1.2.0) (1.0.1)
Requirement already satisfied: future>=0.15.0 in /usr/local/lib/python3.7/dist-packages (from skidl==1.2.0) (0.16.0)
Requirement already satisfied: kinparse>=0.1.0 in /usr/local/lib/python3.7/dist-packages (from skidl==1.2.0) (1.1.0)
Requirement already satisfied: graphviz in /usr/local/lib/python3.7/dist-packages (from skidl==1.2.0) (0.10.1)
Requirement already satisfied: pyparsing>=2.1.1 in /usr/local/lib/python3.7/dist-packages (from kinparse>=0.1.0->skidl==1.2.0) (3.0.9)
Requirement already satisfied: hierplace in /usr/local/lib/python3.7/dist-packages (from kinet2pcb->skidl==1.2.0) (1.0.0)
ERROR: No footprint for R/R1. @ [/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882=>/content/<ipython-input-8-db097054a2e6>:10]
ERROR:skidl:No footprint for R/R1. @ [/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882=>/content/<ipython-input-8-db097054a2e6>:10]
ERROR: No footprint for R/R2. @ [/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882=>/content/<ipython-input-8-db097054a2e6>:10]
ERROR:skidl:No footprint for R/R2. @ [/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882=>/content/<ipython-input-8-db097054a2e6>:10]

0 warnings found while generating netlist.
2 errors found while generating netlist.

(export (version D)\n  (design\n    (source "/content/skidl")\n    (date "07/23/2022 11:42 PM")\n    (tool "SKiDL (1.1.0)"))\n  (components\n    (comp (ref R1)\n      (value 1K)\n      (footprint )\n      (fields\n        (field (name F0) R)\n        (field (name Reference) R)\n        (field (name F1) R)\n        (field (name Value) R)\n        (field (name F3) ~)\n        (field (name Datasheet) ~)\n        (field (name F4) "R res resistor")\n        (field (name ki_keywords) "R res resistor")\n        (field (name F5) Resistor)\n        (field (name ki_description) Resistor)\n        (field (name F6) R_*)\n        (field (name ki_fp_filters) R_*))\n      (libsource (lib Device) (part R))\n      (sheetpath (names /top/16974562689290046177) (tstamps /top/16974562689290046177)))\n    (comp (ref R2)\n      (value 1K)\n      (footprint )\n      (fields\n        (field (name F0) R)\n        (field (name Reference) R)\n        (field (name F1) R)\n        (field (name Value) R)\n        (f
```

This experiment shows another possible improvement:  auto source [footprints](https://kicad.github.io/footprints/) if the user doesn't have the appropriate libraries installed.

**Reply by devbisme** [reply]
So this is an error within Jupyter Book?

Also, in your colab output, I'm not sure why there are missing footprint errors. The footprint is clearly specified in the `Part` call as `Resistor_SMD:R_0603_1608Metric`. And why are there two resistors (R1 and R2) when only one part was created? When I ran this code in a terminal it didn't report any errors and there was only one resistor created.

**Comment by shanemmattner** [comment]
I think the error is within `Thebe` and/or `Jupyter book`.  How frequently do you push updates to SKiDL `master`?

We're probably getting the same error for footprints as we did for schematics:  there's no local copy of the footprint library in the `Thebe` machine so SKiDL throws an error.  Can you auto source from the web if a SKiDL user doesn't have KiCAD footprint libraries installed, similar to the change you implemented for schematic libraries?

I'm guessing `R1`/`R2` refers to the pins of part `R`.

**Comment by devbisme** [comment]
SKiDL doesn't really check for footprint libraries. It doesn't complain as long as there is a string in the `footprint` attribute. The `R1` and `R2` are definitely part references. Were there any preceding `Part` declarations in the colab notebook?

**Reply by shanemmattner** [reply]
Ah, I see.  

I think I found the source of those messages.  Colab doesn't reset variables so if I previously declared the resistor without a footprint it will keep spamming the message:
![r_no_footprint](https://user-images.githubusercontent.com/22162612/180671620-6faa9ff0-5855-435a-937c-854f0c0a6b01.gif)

**Comment by shanemmattner** [comment]
I added a [section](https://shanemmattner.github.io/skidl_jupyter_book/chapter2/06_esp32_min.html) on creating an ESP32 development board.  Let me know what you think.

**Reply by 2e0byo** [reply]
```python
 if len(h_pin.nets)> 0:
     pass
 else:
     header_full = False
     pin += h_pin
     break
```

Should probably be:

```python
 if len(h_pin.nets) == 0:
     header_full = False
     pin += h_pin
     break
 ```
 
 As an aside this project is amazing, and the interactive docs only make it more so.

**Reply by shanemmattner** [reply]
Thanks for the feedback @2e0byo .  Instead of using my original sloppy algorithm in `ESP32 Minimum Dev Board` I changed to using the final function for breaking out any IC.  

```python
def breakout_ic(ic, header):
    headers=[header]
    counter = 1
    # Range through IC pins and connect them to a header
    for n in range(1,len(ic.pins)):
        pin = ic[n]
        headers[-1][counter] += pin
        counter += 1
        # Check if we've reached the end of the header
        if counter > len(header.pins):
            counter = 1
            # Add another copy of the header to the list
            headers.append(header.copy())
            # Disconnect all the pins of the new header
            for i in headers[-1].pins:
                i.disconnect()
```

**Reply by 2e0byo** [reply]
Won't that miss a pin off?  `range(1, len(ic.pins)` counts up to one *less* than the last pin.  In the case of the esp32 the last two are both connected to ground, so it doesn't appear.  [If this is deliberate it should be commented]

Suggested slightly different fn:
```python
def breakout_ic(ic, header):
    headers=[header]
    counter = 1
    # Iterate through IC pins in pin order and connect them to a header
    for n in range(1, len(ic.pins) +1): # ic.pins may not be in footprint order
        pin = ic[n]
        headers[-1][counter] += pin
        counter += 1
        # Check if we've reached the end of the header
        if counter > len(header.pins):
            counter = 1
            # Add another copy of the header to the list
            headers.append(header.copy())
            # Disconnect all the pins of the new header
            for i in headers[-1].pins:
                i.disconnect()
```

correcting (?) the iteration, and explaining why we don't just iterate ic.pins directly for newbies like me ;)

**Reply by devbisme** [reply]
First of all, and most importantly, an end-to-end example like this is really good!

I'm building a (sorely neglected) library of SKiDL modules over at [circuitsascode](https://devbisme.github.io/circuitsascode/). I'm hoping to add MCU modules like this one to it.

I do have a couple of small suggestions:
* SKiDL includes a `generate_pcb()` function that will automatically create a KiCad PCB file and run `HierPlace` on it. This would cut out a few manual steps in your tutorial. However, maybe it is better to show those steps so people know how to do them. IDK.
* Since the ESP32-WROOM module has names for all its pins, why not use them instead of the pin numbers? Then you could use something like `esp32["GND"] += GND` and connect *all* the ground pins. (Also, the module has VDD on pin 2 and GND on pins 1, 15, 38, 39. Looks like you swapped them, maybe?)
* You can "simplify" the reset circuit using something like `V3_3 & r_en_pullup & esp32['EN'] & (c_en_filter | switch[1,3] | switch[2,4]) & GND`. This places the resistor between Vdd and the enable pin and then places the capacitor and both switch contacts in parallel between the enable pin and ground. This eliminates lines of code, but I'm not sure if it's clearer or not.

**Reply by devbisme** [reply]
I didn't want to miss out on the fun of writing the `breakout_ic` function, so here's my version:
```python
from skidl import *

esp32 = esp32 = Part('RF_Module','ESP32-WROOM-32', footprint='RF_Module:ESP32-WROOM-32')

# Create a template that can be used to make copies of the breakout header.
header = Part('Connector', 'Conn_01x05_Male', footprint='Connector_PinHeader_2.54mm:PinHeader_1x05_P2.54mm_Vertical', dest=TEMPLATE)

def breakout_ic(ic, header):

	# Create a list for breakout headers.
	headers = []
	
	# Get all the IC pins in order.
	all_ic_pin_nums = list(range(1, len(ic)+1))
	
	# Break the IC pins into small lists that fit an individual header and make connections.
	for ic_pin_nums in [all_ic_pin_nums[i:i+len(header)] for i in range(0,len(ic),len(header))]:
	
		# Get the header pin numbers 1,2,...,N that match up with the current IC pin numbers.
		hdr_pin_nums = list(range(1,len(ic_pin_nums)+1))
		
		# Add a new header to the end of the list using the header template.
		headers.append(header())
		
		# Connect the header at the end of the list to the IC pins. 
		headers[-1][hdr_pin_nums] += ic[ic_pin_nums]

# Add breakout headers to the ESP32.
breakout_ic(esp32, header)

# Print out the connections to see if we did it right.
for pin in esp32:
	print(f"Connection: {pin.net}")
```

**Reply by shanemmattner** [reply]
@2e0byo `correcting (?) the iteration, and explaining why we don't just iterate ic.pins directly for newbies like me ;)`
Iterating over `part.pins` results in calling the pins out of order (that I intended)
```python
for pin in esp32.pins:
    print(str(pin))
```

```
Pin U1/1/GND/POWER-IN
Pin U1/10/IO25/BIDIRECTIONAL
Pin U1/11/IO26/BIDIRECTIONAL
Pin U1/12/IO27/BIDIRECTIONAL
Pin U1/13/IO14/BIDIRECTIONAL
Pin U1/14/IO12/BIDIRECTIONAL
Pin U1/15/GND/PASSIVE
Pin U1/16/IO13/BIDIRECTIONAL
Pin U1/17/SHD/SD2/BIDIRECTIONAL
Pin U1/18/SWP/SD3/BIDIRECTIONAL
Pin U1/19/SCS/CMD/BIDIRECTIONAL
Pin U1/2/VDD/POWER-IN
Pin U1/20/SCK/CLK/BIDIRECTIONAL
Pin U1/21/SDO/SD0/BIDIRECTIONAL
Pin U1/22/SDI/SD1/BIDIRECTIONAL
Pin U1/23/IO15/BIDIRECTIONAL
Pin U1/24/IO2/BIDIRECTIONAL
Pin U1/25/IO0/BIDIRECTIONAL
Pin U1/26/IO4/BIDIRECTIONAL
Pin U1/27/IO16/BIDIRECTIONAL
Pin U1/28/IO17/BIDIRECTIONAL
Pin U1/29/IO5/BIDIRECTIONAL
Pin U1/3/EN/INPUT
Pin U1/30/IO18/BIDIRECTIONAL
Pin U1/31/IO19/BIDIRECTIONAL
Pin U1/32/NC/NO-CONNECT
Pin U1/33/IO21/BIDIRECTIONAL
Pin U1/34/RXD0/IO3/BIDIRECTIONAL
Pin U1/35/TXD0/IO1/BIDIRECTIONAL
Pin U1/36/IO22/BIDIRECTIONAL
Pin U1/37/IO23/BIDIRECTIONAL
Pin U1/38/GND/PASSIVE
Pin U1/39/GND/PASSIVE
Pin U1/4/SENSOR_VP/INPUT
Pin U1/5/SENSOR_VN/INPUT
Pin U1/6/IO34/INPUT
Pin U1/7/IO35/INPUT
Pin U1/8/IO32/BIDIRECTIONAL
Pin U1/9/IO33/BIDIRECTIONAL
```

**Reply by 2e0byo** [reply]
Oh indeed: I added a comment once I figured that out.  But you're still missing a pin, right?

**Comment by shanemmattner** [comment]
@devbisme your `breakout_ic()` code works well but I'm afraid that it's too Pythonic and will confuse people who aren't great programmers (ie <b>me</b>).  I'd assume more SKiDL users will come from an EE background than programming background so it's probably best to keep the circuit code as simple as possible.  

To respond to your other ideas:
- Using `generate_pcb()` is a good idea.  However, it's not working for me with KiCAD 6.  Here's the [bug](https://github.com/devbisme/skidl/issues/155)
- Using names of pins is better.  Also, I didn't know you could assign all the `GND` pins at once, awesome!
- The `chain connecting` of parts is a cool feature but I think it will confuse users.  Maybe we could demonstrate the idea in these `passives` functions:
```python
def r(value, p1, p2, footprint='Resistor_SMD:R_0603_1608Metric'):
    r = Part('Device', 'R', value=value, footprint=footprint)
    p1 & r & p2
```

**Reply by devbisme** [reply]
> so it's probably best to keep the circuit code as simple as possible.

Agreed, but I would change your code to accept a `Part` template for the `header` argument instead of a fully-instantiated `Part`. Then you can use the template to make headers as needed instead of copying an existing header and then having to disconnect all the pins.

> Using generate_pcb() is a good idea. However, it's not working for me with KiCAD 6.

Thanks, I'll check into it.

> The chain connecting of parts is a cool feature but I think it will confuse users.

Agreed, but I'm not keen on the `passives` functions, either. IMHO, it adds too much baggage. Better to stick with the existing code that uses the `+=` operator.

**Reply by shanemmattner** [reply]
- Good idea on `part templates` for the `breakout_ic()` function
- I'll try using `part templates` instead of the passive functions

**Comment by 2e0byo** [comment]
In the unlikely event the maintainers decide they want #156 in skidl, you can just iterate .pins directly.

**Reply by devbisme** [reply]
Thanks for offering to do this! I made a comment to your PR.

**Reply by 2e0byo** [reply]
For the record in this issue this became #157.  When it gets merged you can do:

```python
for pin in thing.ordered_pins:
     ...
```

I'll not rewrite `breakout_ic` again, but it can be slightly simpler now :D

**Reply by devbisme** [reply]
I pushed the changes for `ordered_pins` to the `development` branch.
